src/hooks.rs

```diff
@@ -0,0 +1,145 @@
+use crate::agent_utils::{TraceresultS, ray_scan, world_pos};
+use crate::err::Error;
+use crate::sdl::SDL_event;
+
+use libc::{RTLD_LAZY, c_char, c_int, dl_iterate_phdr, dl_phdr_info, dlopen, dlsym, size_t};
+use std::ffi::c_void;
+use std::mem::transmute;
+use std::ptr::read_unaligned;
+
+type SdlPushEventFn = unsafe extern "C" fn(*mut SDL_event) -> i32;
+type SdlGetMouseStateFn = unsafe extern "C" fn(*const i32, *const i32) -> u32;
+type SdlGLSwapWindowInnerFn = unsafe extern "C" fn(*const c_void);
+
+type TracelineFn = unsafe extern "C" fn(world_pos, world_pos, u64, bool, *const TraceresultS);
+type IsVisibleFn = unsafe extern "C" fn(world_pos, world_pos, u64, bool) -> bool;
+
+pub static mut SDL_PUSHEVENT: Option<SdlPushEventFn> = None;
+pub static mut SDL_GETMOUSESTATE: Option<SdlGetMouseStateFn> = None;
+pub static mut TRACE_LINE_FUNC: Option<TracelineFn> = None;
+pub static mut IS_VISIBLE_FUNC: Option<IsVisibleFn> = None;
+
+static mut MUTABLE_INNER_FUNC_PTR: Option<*mut unsafe extern "C" fn(*const c_void)> = None;
+static mut HOOK_ORIGINAL_INNER_FUNC: Option<SdlGLSwapWindowInnerFn> = None;
+
+macro_rules! cstr_static {
+    ($s:expr) => {
+        concat!($s, "\0").as_ptr() as *const c_char
+    };
+}
+
+unsafe extern "C" fn hook_func(window: *const c_void) {
+    ray_scan(2, 0.0, 360.0).expect("ray tracing error");
+
+    unsafe {
+        match HOOK_ORIGINAL_INNER_FUNC {
+            Some(func) => func(window),
+            None => (),
+        }
+    }
+}
+
+pub fn sdl_gl_swap_window_hook(sdl_gl_swap_window_handle: *mut c_void) -> Result<(), Error> {
+    unsafe {
+        let wrapper_offset_location = sdl_gl_swap_window_handle as u64 + 0x4 + 0x2;
+
+        let offset = read_unaligned(wrapper_offset_location as *const u32);
```

Comment:
For functions like `transmute` and `read_unaligned`, or anything thats not defined by you, its more proper to include at least the parent mod in the call: 
```suggestion
        let offset = ptr::read_unaligned(wrapper_offset_location as *const u32);
```